cwms-data-api/src/main/java/cwms/cda/data/dao/location/kind/OutletDao.java

```diff
@@ -0,0 +1,240 @@
+import cwms.cda.data.dto.LocationGroup;
+import cwms.cda.data.dto.location.kind.CompoundOutletRecord;
+import cwms.cda.data.dto.location.kind.Outlet;
+import cwms.cda.data.dto.location.kind.ProjectStructure;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.jooq.Configuration;
+import org.jooq.DSLContext;
+import org.jooq.exception.DataAccessException;
+import org.jooq.impl.DSL;
+import usace.cwms.db.dao.util.OracleTypeMap;
+import usace.cwms.db.jooq.codegen.packages.CWMS_OUTLET_PACKAGE;
+import usace.cwms.db.jooq.codegen.udt.records.LOCATION_REF_T;
+import usace.cwms.db.jooq.codegen.udt.records.PROJECT_STRUCTURE_OBJ_T;
+import usace.cwms.db.jooq.codegen.udt.records.STR_TAB_T;
+import usace.cwms.db.jooq.codegen.udt.records.STR_TAB_TAB_T;
+
+public class OutletDao extends JooqDao<Outlet> {
+    private static final FluentLogger LOGGER = FluentLogger.forEnclosingClass();
+    private static final String RATING_LOC_GROUP_CATEGORY = "Rating";
+
+    public OutletDao(DSLContext dsl) {
+        super(dsl);
+    }
+
+    public List<Outlet> retrieveOutletsForProject(String projectId, String officeId) {
+        return connectionResult(dsl, conn -> {
+            Configuration config = DSL.using(conn).configuration();
+            Map<String, List<CompoundOutletRecord>> compoundOutlets = retrieveCompoundOutletsForProject(config,
+                                                                                                        projectId,
+                                                                                                        officeId);
+            LOCATION_REF_T locRef = LocationUtil.getLocationRef(projectId, officeId);
+
+            LocationGroupDao locGroupDao = new LocationGroupDao(dsl);
+            List<LocationGroup> groups = locGroupDao.getLocationGroups(officeId, RATING_LOC_GROUP_CATEGORY);
+
+            return CWMS_OUTLET_PACKAGE.call_RETRIEVE_OUTLETS(config, locRef)
+                                      .stream()
+                                      .map(struct -> mapToOutlet(struct, compoundOutlets, groups))
+                                      .collect(Collectors.toList());
+        });
+    }
+
+    public Outlet retrieveOutlet(String locationId, String officeId) {
+        return connectionResult(dsl, conn -> {
+            LOCATION_REF_T locRef = LocationUtil.getLocationRef(locationId, officeId);
+            Configuration config = DSL.using(conn).configuration();
+            PROJECT_STRUCTURE_OBJ_T outletStruct = CWMS_OUTLET_PACKAGE.call_RETRIEVE_OUTLET(config, locRef);
+
+            CwmsId projectId = LocationUtil.getLocationIdentifier(outletStruct.getPROJECT_LOCATION_REF());
+
+            Map<String, List<CompoundOutletRecord>> records = new HashMap<>();
+            try {
+                List<CompoundOutletRecord> compoundOutletRecords = retrieveCompoundOutlet(config, locationId,
```

Comment:
For every outlet, two DB calls are made?

Reply:
I'll talk on the one about 3 db calls